src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs

```diff
@@ -0,0 +1,187 @@
+
+            PSObject result = new PSObject();
+
+            result.Properties.Add(new PSNoteProperty("Owner", owner));
+            result.Properties.Add(new PSNoteProperty("Group", group));
+            result.Properties.Add(new PSNoteProperty("DiscretionaryAcl", discretionaryAcl));
+            result.Properties.Add(new PSNoteProperty("SystemAcl", systemAcl));
+
+            WriteObject(result);
+        }
+
+        /// <summary>
+        /// Translates a SID into a NT Account
+        /// </summary>
+        private string ConvertToNtAccount(SecurityIdentifier sid)
+        {
+            try
+            {
+                return sid.Translate(typeof(NTAccount)).ToString();
+            }
+            catch {}
+
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the access rights that apply to an access mask, preferring right types
+        /// of 'Type' if specified.
+        /// </summary>
+        private List<string> GetAccessRights(long accessMask, string type)
+        {
+            IDictionary<string, Type> rightTypes = new Dictionary<string, Type>
+            {
+                { "FileSystemRights", typeof(System.Security.AccessControl.FileSystemRights) },
+                { "RegistryRights", typeof(System.Security.AccessControl.RegistryRights) },
+#if !CORECLR
+                { "ActiveDirectoryRights", typeof(System.DirectoryServices.ActiveDirectoryRights) },
+#endif
+                { "MutexRights", typeof(System.Security.AccessControl.MutexRights) },
+                { "SemaphoreRights", typeof(System.Security.AccessControl.SemaphoreRights) },
+#if !CORECLR
+                { "CryptoKeyRights", typeof(System.Security.AccessControl.CryptoKeyRights) },
+#endif
+                { "EventWaitHandleRights", typeof(System.Security.AccessControl.EventWaitHandleRights) }
+            };
+
+            List<Type> typesToExamine = rightTypes.Values.ToList();
+            
+            // If they know the access mask represents a certain type, prefer its names
+            // (i.e.: CreateLink for the registry over CreateDirectories for the filesystem)
+            if (!String.IsNullOrEmpty(type))
+            {
+                typesToExamine.Insert(0, rightTypes[type]);
+            }
+            
+            // Stores the access types we've found that apply
+            List<string> foundAccess = new List<string>();
+
+            // Store the access types we've already seen, so that we don't report access
+            // flags that are essentially duplicate. Many of the access values in the different
+            // enumerations have the same value but with different names.
+            HashSet<long> foundValues = new HashSet<long>();
+
+            // Go through the entries in the different right types, and see if they apply to the
+            // provided access mask. If they do, then add that to the result.
+            foreach (var rightType in typesToExamine)
```

Comment:
I re-read the document and made any tests - My conclusion is full discard the algorithm:

1. From the [docs](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379557%28v=vs.85%29.aspx):
> Each type of securable object defines its own set of specific access rights and its own mapping of generic access rights. For information about the specific and generic access rights for each type of securable object, see the overview for that type of object.

- the conclusion is  an SDDL has no meaning without object entity - the cmdlet should get the object type as argument.

2. The cmdlet calls  [Security.AccessControl.CommonSecurityDescriptor]::new($false,$false,$Sddl)
[The constructor](https://github.com/dotnet/corefx/blob/0e4095bc5da407df6cbe0aa350e1b2ccf32696a4/src/System.Security.AccessControl/src/System/Security/AccessControl/SecurityDescriptor.cs#L949) in Core FX

I tested the cmdlet with (1)  [Security.AccessControl.CommonSecurityDescriptor]::new($false,$false,$Sddl) and (2)  [Security.AccessControl.CommonSecurityDescriptor]::new($true,$false,$Sddl) for path "c:\test" (it is container) and got different results.

- the conclusion is the cmdlet should take into account `isContainer` (folders/containers) and `isDS` (directory objects and containers).

3. CoreFX can throw exceptions (ArgumentException) - maybe the cmdlet should catch them and re-throw (put CoreFX exception in InnerException)

4.  The cmdlet should have strongly typed output. In current PSCustomObject we should add "Type".

5. "Type" - we can define it as Enum ( remove unsupported member by #if !Unix ) and exclude ValidateSet at all.

6. We should add tests for all object types.